src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java

```diff
@@ -0,0 +1,436 @@
+    }
+  }
+
+  private TiTableInfo tableInfo;
+  private TiIndexInfo indexInfo;
+  private final List<TiColumnRef> fields = new ArrayList<>();
+  private final List<TiExpr> where = new ArrayList<>();
+  private final List<TiByItem> groupByItems = new ArrayList<>();
+  private final List<TiByItem> orderByItems = new ArrayList<>();
+  // System like Spark has different type promotion rules
+  // we need a cast to target when given
+  private final List<Pair<TiExpr, DataType>> aggregates = new ArrayList<>();
+  private final List<Coprocessor.KeyRange> keyRanges = new ArrayList<>();
+
+  private int limit;
+  private int timeZoneOffset;
+  private long flags;
+  private long startTs;
+  private TiExpr having;
+  private boolean distinct;
+
+  public void bind() {
+    getFields().forEach(expr -> expr.bind(tableInfo));
+    getWhere().forEach(expr -> expr.bind(tableInfo));
+    getGroupByItems().forEach(item -> item.getExpr().bind(tableInfo));
+    getOrderByItems().forEach(item -> item.getExpr().bind(tableInfo));
+    getAggregates().forEach(expr -> expr.bind(tableInfo));
+    if (having != null) {
+      having.bind(tableInfo);
+    }
+  }
+
+  public DAGRequest buildScan(boolean idxScan) {
+    if (idxScan) {
+      return buildIndexScan();
+    } else {
+      return buildTableScan();
+    }
+  }
+
+  // See TiDB source code: executor/builder.go:945
+  private DAGRequest buildIndexScan() {
+    checkArgument(startTs != 0, "timestamp is 0");
+    if (indexInfo == null) {
+      throw new TiClientInternalException("Index is empty for index scan");
+    }
+    DAGRequest.Builder dagRequestBuilder = DAGRequest.newBuilder();
+    Executor.Builder executorBuilder = Executor.newBuilder();
+    IndexScan.Builder indexScanBuilder = IndexScan.newBuilder();
+    indexScanBuilder
+            .setTableId(tableInfo.getId())
+            .setIndexId(indexInfo.getId());
+    dagRequestBuilder.addExecutors(executorBuilder.setIdxScan(indexScanBuilder));
+
+    return dagRequestBuilder
+            .setFlags(flags)
+            .setTimeZoneOffset(timeZoneOffset)
+            .setStartTs(startTs)
+            .build();
+  }
+
+  // See TiDB source code: executor/builder.go:890
+  private DAGRequest buildTableScan() {
+    checkArgument(startTs != 0, "timestamp is 0");
+    DAGRequest.Builder dagRequestBuilder = DAGRequest.newBuilder();
+    Executor.Builder executorBuilder = Executor.newBuilder();
+    TableScan.Builder tblScanBuilder = TableScan.newBuilder();
+
+
+    // Step1. Add columns to first executor
+    tableInfo.getColumns().forEach(tiColumnInfo -> tblScanBuilder.addColumns(tiColumnInfo.toProto(tableInfo)));
+    executorBuilder.setTp(ExecType.TypeTableScan);
+    tblScanBuilder.setTableId(tableInfo.getId());
+    dagRequestBuilder.addExecutors(executorBuilder.setTblScan(tblScanBuilder));
+    executorBuilder.clear();
+
+    // Step2. Add others
+    // DO NOT EDIT EXPRESSION CONSTRUCTION ORDER
+    // Or make sure the construction order is below:
```

Comment:
It is better we have a function `validate`. We can validate expression once we create it.

Reply:
Agree, working on it.

Reply:
Done.